client/src/app/template-questionnaire/answer-key.json

```diff
@@ -110,19 +110,22 @@
     ]
   },
   "preNatalQuantity": {
-    "Yes, All": [
+    "Yes, Some": [
       {
         "schemaName": "donor_organism",
-        "category": "biomaterial"
+        "category": "biomaterial",
+        "includeModules": [
+          "organism_age_unit",
```

Comment:
`organism_age` and `organism_age_unit` are two separate modules.

Same with `gestational_age` and `gestational_age_unit`

Reply:
What Marion told me is that `organism_age_unit` is a an ontology module and the corresponding `organism_age` is a field. This is the same for `gestational_age_unit` and `gestational_age`.

Reply:
Ahh I see, but in that case we'll need "includeFields". The spreadsheet generator service will only include what's specified in these type specs, be it fields or modules. This is why I originally wrote it to be explicit exclude rather than explicit include. Functionally, it'll work the same if you just put `organism_age` into the `includeModules` array (maybe we should rename includeModules to includeFields, a more generic term).

But yes, everything that's included for a TypeSchema needs to be explicitly included. The spreadsheet generator service doesn't have built in assumptions about what to specifically include.

Reply:
Ok. So, putting these fields in the `includeModules` property should do it then?

Reply:
Yep, that should work

Reply:
@rdgoite All fields need to be explicitly included. I believe in Marion's specification doc she mentions that all fields need to be included by default. Those need to be specified throughout

Reply:
Yes, I have those default fields already. See `default_type_specs` in `template-questionnaire.data.ts`.

Reply:
I think those are the default "modules". The doc mentions ".. In general we assume that all properties in the ‘core’ module for each schema is included as well as _any properties_ that are defined directly in the type schema..."

"any properties" refers to non-module fields that are outside of this questionnaire. e.g `biomaterial_core`, `sex`, `is_living` (see https://schema.dev.archive.data.humancellatlas.org/type/biomaterial/15.5.0/donor_organism). These fields need to be specified, given that we're going for explicit include.

Reply:
My assumption was that the template generator does this already. My understanding was that whatever we sent from the template questionnaire are additional information on top of the defaults. I thought the exclusion was hard to make work because the UI would then have to know the ins and outs of the metadata schema to make the decision.

Reply:
The idea was to maintain knowledge about interpreting the questionnaire solely UI, while the spreadsheet generator service is capable of generating spreadsheets, regardless of any particular questionnaire. Putting those "defaults" in the questionnaire would mean that both the UI and generator service would need to be changed as the questionnaire requirements change. 

Keeping this in the UI means we only have to change the implementation in one place, and also leaves the spreadsheet generator service powerful enough to generate any kind of spreadsheet, rather than just those that fit the criteria of this specific questionnaire/use-case

Reply:
There are decisions that we can only make in the UI, yes, but even at the current state of the code, we seem to be leaking a significant portion of business decision in the UI here. In general, the less the client needs to send to achieve what it wants to do, the better; I think this makes for a better user experience. I don't think there's harm in having sensible defaults in the template generator. If we ever decide to open up the backend API for this, we'd probably want programmatic users to be able to POST an empty JSON and still get something sensible; a blank spreadsheet would too little to be useful, but then a fully loaded spreadsheet would be too much. In the case of the latter, it also puts the onus on the client to understand what it does *not* want.

Just so that we get something out quickly, I'm going to add those fields in the UI for now. However, we probably need to rethink how we do things with template generation.

Reply:
We could have a separate API that's aware of the use case of this particular questionnaire's requirements for spreadsheet generation, and then that could forward a call, paramterised with defaults, to the existing spreadsheet generator API. This could live in the ingest-broker or the UI's NodeJS backend if for some reason we don't want it defined with the Angular app's code, though I think it's sensible for it to live there since it is logic pertaining to that particular application of spreadsheet generation. The "UI" isn't just a "view", it's also a full blown application and it should contain business logic pertinent to the interpretation of GUI interactions e.g forms/questionnaires.

Reply:
Don't you think that's a lot more work than it needs to be? Anyway, we don't have to make a decision about this right now, so let's just revisit this.

Reply:
No, I think the path of least code and complexity is to have the defaults configured in the angular app. 

As an aside on what I think the best architecture is: The app is already aware of the specific inputs that require default parameters. To be more general, the app is already aware of the questionnaire business use-case and the pertaining logic.

Reply:
I think the schema template service will have the knowledge which will be the default modules to be included by checking the required fields in a schema of concrete type (donor). Core module and the required fields which corresponds to a module schema should be the default.

Reply:
There are "default" fields that aren't marked as required, e.g `genus_species`. 

We could define `autoIncludeRequiredFields: boolean` in TypeSpec. That would require less lines in the definition of what should be included by default. Fundamentally though, it doesn't really reduce the complexity of this problem; we would still have had to examine the relevant schemas for which fields to include as "default"

Reply:
Interesting, i wonder what's the basis for a module to be in the default. In that case, i think it's alright to just maintain the list from the wranglers. But the question is, where should that default list live? UI or Schema template service?

Reply:
> Interesting, i wonder what's the basis for a module to be in the default

I think it's just what the wranglers decided to be default.

I think the list should live next to its application e.g the UI. We could have another "convenience" API that maintains the list and passes the default values to the existing unopinionated spreadsheet generation API. If the schemas themselves supported "default", alongside "required", then it would be sensible for the existing generator service to handle these defaults.